examples/experimental/dagster-airlift/dagster_airlift/core/state_backed_defs_loader.py

```diff
@@ -0,0 +1,127 @@
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from typing import Any, Generator, Generic, TypeVar
+
+from dagster import Definitions
+from dagster._core.definitions.definitions_loader import DefinitionsLoadContext, DefinitionsLoadType
+from dagster._core.definitions.metadata.metadata_value import (
+    CodeLocationReconstructionMetadataValue,
+)
+from dagster._core.definitions.repository_definition.repository_definition import RepositoryLoadData
+
+TState = TypeVar("TState")
+
+
+class StateBackedDefinitionsLoader(ABC, Generic[TState]):
+    """A class for building state-backed definitions. In a structured way. It
+    handles manipulation of the DefinitionsLoadContext on the user's behalf.
+
+    The goal here is so that an unreliable backing source is fetched only
+    at code server load time, which defines the code location. When, for example,
+    a run worker is launched, the backing source is not queryed again, and
+    only `defs_from_state` is called.
+
+    Current meant for internal usage only hence TState must be a marked as whitelist_for_serdes.
+
+    Args:
+        defs_key (str): The unique key for the definitions. Must be unique per code location.
+
+    Examples:
+
+    .. code-block:: python
+
+        @whitelist_for_serdes
+        @record
+        class ExampleDefState:
+            a_string: str
+
+        class ExampleStateBackedDefinitionsLoader(StateBackedDefinitionsLoader[ExampleDefState]):
+            def fetch_backing_state(self) -> ExampleDefState:
+                # Fetch from potentially unreliable API (e.g. Rest API)
+                return ExampleDefState(a_string="foo")
+
+            def defs_from_state(self, state: ExampleDefState) -> Definitions:
+                # Construct or reconstruct the Definitions from the previously
+                # fetched state.
+                return Definitions([AssetSpec(key=state.a_string)])
+    """
+
+    @property
+    @abstractmethod
+    def defs_key(self) -> str:
+        """The unique key for the definitions. Must be unique per code location."""
+        ...
+
+    @abstractmethod
+    def fetch_backing_state(self) -> TState:
```

Comment:
I would prefer `fetch_state`-- "backing" qualifier seems unnecessary and potentially confusing here in light of `defs_from_state`